app/src/main/java/com/github/sdpcoachme/location/provider/FusedLocationProvider.kt

```diff
@@ -0,0 +1,184 @@
+     * outputs if the user has granted it or not. See init(...).
+     */
+    private lateinit var requestPermissionLauncher: ActivityResultLauncher<String>
+
+    /**
+     * Create an activity for result to display a window to request location setting.
+     * The contract is a predefined "function" which takes an IntentSenderRequest as input and
+     * outputs if the user has enabled the setting or not. See init(...).
+     */
+    private lateinit var requestSettingLauncher:  ActivityResultLauncher<IntentSenderRequest>
+
+    override fun init(context: ComponentActivity, userInfo: CompletableFuture<UserInfo>) {
+        appContext = context
+        user = userInfo
+        lastUserLocation = mutableStateOf(null)
+        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(appContext)
+
+        requestPermissionLauncher = appContext.registerForActivityResult(
+            ActivityResultContracts.RequestPermission()
+        ) { isGranted ->
+            if (isGranted) {
+                checkLocationSetting()
+            } else {
+                setLocationToAddress()
+            }
+        }
+
+        requestSettingLauncher = appContext.registerForActivityResult(
+            ActivityResultContracts.StartIntentSenderForResult()
+        ) {
+            if (it.resultCode == RESULT_OK) {
+                // User enabled device location, we can now retrieve it
+                getDeviceLocation(0)
+            } else {
+                // User did not enable device location
+                setLocationToAddress()
+            }
+        }
+    }
+
+    override fun requestPermission() {
+        requestPermissionLauncher.launch(ACCESS_FINE_LOCATION)
+    }
+
+    /**
+     * Especially useful after application installation.
+     */
+    override fun locationIsPermitted(): Boolean {
+        val locationPermission = ContextCompat.checkSelfPermission(appContext, ACCESS_FINE_LOCATION)
+        return locationPermission == PERMISSION_GRANTED
+    }
+
+    /**
+     * This function is always called with granted permissions
+     * and the device location settings enabled. This safely allows the recursive call needed in the
+     * case the device location service takes time to be deployed. A max iteration "timeout" has
+     * been set to ensure termination.
+     *
+     * Notice that the fusedLocationProviderClient.lastLocation task is null if the location is
+     * disabled on the device (also even if the last location was previously retrieved because
+     * disabling the device location clears the cache).
+     */
+    @SuppressLint("MissingPermission") //permission is checked before the call
+    private fun getDeviceLocation(delay: Long) {
+        // getDeviceLocation should not be called more than MAX_LOCATION_DELAY time
+        if (delay >= DELAY) {
+            error("getDeviceLocation has reached its max recursive delay")
```

Comment:
I have noticed that the termination is based on this delay, could you explain how it was chosen?
Also, the documentation could be updated above.

Reply:
The DELAY has been set to 5000 iterations after several local tests. `getDeviceLocation(...)` should normally not be called over around 3000 times. If it happens, delay should be adjusted or another implementation found.